lib/zip/file.rb

```diff
@@ -405,14 +405,13 @@ def check_file(path)
     def on_success_replace
       tmpfile      = get_tempfile
       tmp_filename = tmpfile.path
-      ObjectSpace.undefine_finalizer(tmpfile)
-      tmpfile.close
       if yield tmp_filename
-        ::File.rename(tmp_filename, name)
+        new_file = ::File.new(name, "w")
+        new_file.write tmpfile.read
```

Comment:
The tempfile stores all changes prior to `commit` or `close`.  Its size is unbounded so it shouldn't be loaded into memory all at once (#233, #250).